Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -0,0 +1,448 @@
+power series over a complete local ring, such that
+the image of `g` in the residue field is not zero. Let `n` be the order of the image of `g` in the
+residue field. Then there exists a power series `q` and a polynomial `r` of degree `< n`, such that
+`f = g * q + r`. -/
+theorem exists_isWeierstrassDivision [IsAdicComplete (IsLocalRing.maximalIdeal A) A]
+    (hg : g.map (IsLocalRing.residue A) ≠ 0) : ∃ q r, f.IsWeierstrassDivision g q r :=
+  ⟨_, _, (IsWeierstrassDivisor.of_map_ne_zero hg).isWeierstrassDivisionAt_div_mod f⟩
+
+-- Unfortunately there is no Unicode subscript `w`.
+
+/-- The `q` in Weierstrass division, denoted by `f /ʷ g`. Note that when the image of `g` in the
+residue field is zero, this is defined to be zero. -/
+noncomputable def weierstrassDiv [IsPrecomplete (IsLocalRing.maximalIdeal A) A] : A⟦X⟧ :=
+  open scoped Classical in
+  if hg : g.map (IsLocalRing.residue A) ≠ 0 then
+    (IsWeierstrassDivisor.of_map_ne_zero hg).div f
+  else
+    0
+
+/-- The `r` in Weierstrass division, denoted by `f %ʷ g`. Note that when the image of `g` in the
+residue field is zero, this is defined to be zero. -/
+noncomputable def weierstrassMod [IsPrecomplete (IsLocalRing.maximalIdeal A) A] : A[X] :=
+  open scoped Classical in
+  if hg : g.map (IsLocalRing.residue A) ≠ 0 then
+    (IsWeierstrassDivisor.of_map_ne_zero hg).mod f
+  else
+    0
+
+@[inherit_doc]
+infixl:70 " /ʷ " => weierstrassDiv
+
+@[inherit_doc]
+infixl:70 " %ʷ " => weierstrassMod
+
+@[simp]
+theorem weierstrassDiv_zero_right [IsPrecomplete (IsLocalRing.maximalIdeal A) A] : f /ʷ 0 = 0 := by
+  rw [weierstrassDiv, dif_neg (by simp)]
+
+@[simp]
+theorem weierstrassMod_zero_right [IsPrecomplete (IsLocalRing.maximalIdeal A) A] : f %ʷ 0 = 0 := by
+  rw [weierstrassMod, dif_neg (by simp)]
+
+theorem degree_weierstrassMod_lt [IsPrecomplete (IsLocalRing.maximalIdeal A) A] :
+    (f %ʷ g).degree < (g.map (IsLocalRing.residue A)).order.toNat := by
+  rw [weierstrassMod]
+  split_ifs with hg
+  · exact degree_trunc_lt _ _
+  · nontriviality A
+    rw [Polynomial.degree_zero]
+    exact WithBot.bot_lt_coe _
+
+section
+
+variable {g} (hg : g.map (IsLocalRing.residue A) ≠ 0)
+include hg
+
+theorem isWeierstrassDivision_weierstrassDiv_weierstrassMod
+    [IsAdicComplete (IsLocalRing.maximalIdeal A) A] :
+    f.IsWeierstrassDivision g (f /ʷ g) (f %ʷ g) := by
+  simp_rw [weierstrassDiv, weierstrassMod, dif_pos hg]
+  exact (IsWeierstrassDivisor.of_map_ne_zero hg).isWeierstrassDivisionAt_div_mod f
+
+theorem eq_mul_weierstrassDiv_add_weierstrassMod
+    [IsAdicComplete (IsLocalRing.maximalIdeal A) A] :
+    f = g * (f /ʷ g) + (f %ʷ g) := by
+  simp_rw [weierstrassDiv, weierstrassMod, dif_pos hg]
+  exact ((IsWeierstrassDivisor.of_map_ne_zero hg).isWeierstrassDivisionAt_div_mod f).2
+
+/-- The quotient `q` and remainder `r` in the Weierstrass division `0 / g` are equal to `0`. -/
+theorem eq_zero_of_weierstrass_division [IsHausdorff (IsLocalRing.maximalIdeal A) A]
```

Comment:
The docstring should go to `weierstrassDiv_zero_left`, not here (using `/ʷ`), not here. Also the name should be changed, there is no weierstrass_division in the statement.

Reply:
This remove is removed (as well as the following one), in favor of the result already stated for `IsWeierstrassDivisorAt`.